// See LICENSE for licensing terms.
// Protocol Buffers extensions for defining auto-generateable validators for messages.
syntax = "proto3";
package validator;

import "google/protobuf/descriptor.proto";

// TODO(vici) google not support to regist a unique extension Id for new plugin from now on.
extend google.protobuf.FieldOptions {
  FieldValidator field = 1059;
}

message FieldValidator {
  // Uses a Golang RE2-syntax regex to match the field contents.
  string regex = 1;
  // Field value of integer strictly greater than this value.
  int64 int_gt = 2;
  // Field value of integer strictly smaller than this value.
  int64 int_lt = 3;
  int64 int_gte = 4;
  int64 int_lte = 5;
  double float_gt = 6;
  // Floating-point value compared to which the field content should be smaller or equal.
  double float_lt = 7;
  double float_gte = 8;
  // Floating-point value compared to which the field content should be smaller or equal.
  double float_lte = 9;
  // Used for nested message types, requires that the message type exists.
  bool msg_exists = 10;
  // Human error specifies a user-customizable error that is visible to the user.
  string human_error = 11;
  bool string_not_empty = 12;
  // Repeated field with at least this number of elements.
  int64 repeated_count_min = 13;
  // Repeated field with at most this number of elements.
  int64 repeated_count_max = 14;
  // Field value of length greater than this value.
  int64 length_gt = 15;
  // Field value of length smaller than this value.
  int64 length_lt = 16;
  // Field value of integer strictly equal this value.
  int64 length_eq = 17;

}
